File: onedal/interop/homogen_table.py
# ==============================================================================
# Copyright 2023 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import onedal
import onedal.interop.buffer as buffer
import onedal.interop.dlpack as dlpack
import onedal.interop.sua as sua
from onedal.interop.utils import is_host_policy

from .array import is_array_entity, to_array

table = onedal._backend.data_management.table
homogen_table = onedal._backend.data_management.homogen_table
homogen_kind = onedal._backend.data_management.table_kind.homogen


def is_native_homogen(entity) -> bool:
    if isinstance(entity, table):
        kind = entity.get_kind()
        return kind == homogen_kind
    return isinstance(entity, homogen_table)


def is_python_homogen(entity) -> bool:
    conditions = [sua.is_sua_table, dlpack.is_dlpack_table, buffer.is_buffer_table]
    return any(map(lambda check: check(entity), conditions))


def is_homogen_entity(entity) -> bool:
    conditions = [is_native_homogen, is_python_homogen, is_array_entity]
    return any(map(lambda check: check(entity), conditions))


def to_homogen_table_array(entity) -> homogen_table:
    assert is_array_entity(entity)
    entity = to_array(entity)
    count = entity.get_count()
    return homogen_table(entity, count, 1)


def to_homogen_table_native(entity) -> homogen_table:
    assert is_native_homogen(entity)
    return homogen_table(entity)


def to_homogen_table_python(entity) -> homogen_table:
    assert is_python_homogen(entity)
    if sua.is_sua_table(entity):
        return sua.to_homogen_table(entity)
    elif dlpack.is_dlpack_table(entity):
        return dlpack.to_homogen_table(entity)
    elif buffer.is_buffer_table(entity):
        return buffer.to_homogen_table(entity)
    else:
        raise ValueError("Not a python homogen table")


def to_homogen_table(entity) -> homogen_table:
    assert is_homogen_entity(entity)
    if is_array_entity(entity):
        result = to_homogen_table_array(entity)
    elif is_native_homogen(entity):
        result = to_homogen_table_native(entity)
    elif is_python_homogen(entity):
        result = to_homogen_table_python(entity)
    else:
        raise ValueError("Not a homogen table")
    assert is_native_homogen(result)
    return result


def from_homogen_table_native(table):
    assert is_native_homogen(table)
    data = table.get_data()
    policy = data.get_policy()
    device = policy.get_device_name()
    if is_host_policy(policy):
        return buffer.from_homogen_table(table)
    elif device in ["cpu", "gpu"]:
        return sua.from_homogen_table(table)
    else:
        raise ValueError("unknown device")


def from_homogen_table_python(table):
    assert is_python_homogen(table)
    return table


def from_homogen_table(table):
    assert is_homogen_entity(table)
    if is_native_homogen(table):
        homogen = homogen_table(table)
        result = from_homogen_table_native(homogen)
    elif is_python_homogen(table):
        result = from_homogen_table_python(table)
    else:
        raise ValueError("Not able to convert from homogen table")
    assert is_python_homogen(result)
    return result

File: onedal/data_management/homogen_table/homogen_table.cpp
/*******************************************************************************
* Copyright 2024 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

#include <pybind11/pybind11.h>

#include "onedal/common.hpp"
#include "onedal/common/dtype_dispatcher.hpp"
#include "onedal/data_management/table/table_iface.hpp"

#include "oneapi/dal/array.hpp"
#include "oneapi/dal/common.hpp"
#include "oneapi/dal/detail/common.hpp"

#include "oneapi/dal/table/common.hpp"
#include "oneapi/dal/table/homogen.hpp"
#include "oneapi/dal/table/detail/table_utils.hpp"

#include "oneapi/dal/detail/common.hpp"

namespace py = pybind11;

namespace oneapi::dal::python::data_management {

template <typename Type, typename Table = dal::homogen_table>
inline void instantiate_homogen_constructor_impl(py::class_<Table>& py_table) {
    py_table.def(py::init([](const array<Type>& arr, std::int64_t rc, std::int64_t cc) {
        return homogen_table::wrap<Type>(arr, rc, cc);
    }));
    py_table.def(
        py::init([](const array<Type>& arr, std::int64_t rc, std::int64_t cc, data_layout dl) {
            return homogen_table::wrap<Type>(arr, rc, cc, dl);
        }));
}

template <typename Table, typename... Types>
inline void instantiate_homogen_constructor(py::class_<Table>& py_table,
                                            const std::tuple<Types...>* const = nullptr) {
    static_assert(std::is_same_v<Table, homogen_table>);
    return detail::apply(
        [&](auto type_tag) -> void {
            using type_t = std::decay_t<decltype(type_tag)>;
            instantiate_homogen_constructor_impl<type_t>(py_table);
        },
        Types{}...);
}

inline std::int64_t get_count(const dal::homogen_table& t) {
    const std::int64_t row_count = t.get_row_count();
    const std::int64_t col_count = t.get_column_count();
    return detail::check_mul_overflow(row_count, col_count);
}

template <typename Type>
dal::array<Type> get_data(const dal::homogen_table& table) {
    const std::int64_t elem_size = //
        detail::integral_cast<std::int64_t>(sizeof(Type));

    auto impl = detail::get_homogen_table_iface(table);
    const std::int64_t count = get_count(table);

    dal::array<dal::byte_t> data = impl->get_data();

    const std::int64_t size = //
        detail::check_mul_overflow(count, elem_size);

    if (size != data.get_count()) {
        throw std::length_error("Incorrect data size");
    }

    if (data.has_mutable_data()) {
        dal::byte_t* raw_ptr = data.get_mutable_data();
        Type* ptr = reinterpret_cast<Type*>(raw_ptr);
        return dal::array<Type>(data, ptr, count);
    }
    else {
        const dal::byte_t* raw_ptr = data.get_data();
        const Type* ptr = reinterpret_cast<const Type*>(raw_ptr);
        return dal::array<Type>(data, ptr, count);
    }
}

py::object get_data_array(const dal::homogen_table& table) {
    const table_metadata& meta = table.get_metadata();
    const data_type dtype = meta.get_data_type(0l);

    return detail::dispatch_by_data_type(dtype, [&](auto type_tag) -> py::object {
        using type_t = std::decay_t<decltype(type_tag)>;
        auto array = get_data<type_t>(table);
        return py::cast(std::move(array));
    });
}

void instantiate_homogen_table(py::module& pm) {
    constexpr const char name[] = "homogen_table";
    py::class_<dal::homogen_table> py_homogen_table(pm, name);

    py_homogen_table.def(py::init<dal::table>());
    py_homogen_table.def("get_data", &get_data_array);

    instantiate_table_iface(py_homogen_table);

    constexpr const supported_types_t* types = nullptr;
    instantiate_homogen_constructor(py_homogen_table, types);
}

} // namespace oneapi::dal::python::data_management
